File: 121.java
// LeetCode 121 Best time to buy and Sell stock
// Reference: https://www.youtube.com/watch?v=E2-heUEnZKU

class Solution {
    public int maxProfit(int[] prices) {
        // at the beginning, the minimum is the first price
        int buy_price = prices[0];

        // at the beginning, the profit is 0
        int profit = 0;

        for (int i = 0; i < prices.length; i++){
            //if the current price is less, update buy_price
            if (buy_price > prices[i]){

                buy_price = prices[i];

            } else {
                //check if we get a better profit
                int current_profit = prices[i] - buy_price;
                profit = Math.max(current_profit, profit);
            }
        }
        return profit;
    }
}

File: 121.js
// LeetCode 121 Best time to buy and Sell stock
// Reference: https://leetcode.com/problems/best-time-to-buy-and-sell-stock/solutions/1735550/python-javascript-easy-solution-with-very-clear-explanation/?envType=study-plan-v2&envId=top-interview-150

/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    let left = 0; // Buy
    let right = 1; // Sell
    let max_profit = 0;
    while (right < prices.length){
        if (prices[left] < prices[right]){
            let profit = prices[right] - prices[left];
            // Our current profit

            max_profit = Math.max(max_profit, profit);
        } else {
            left = right;
        }

        right ++;
    }
    return max_profit;
    
};